firedrake/assign.py

```diff
@@ -0,0 +1,130 @@
+import abc
+
+import numpy as np
+from pyop2.utils import cached_property
+import ufl
+from ufl.algorithms import extract_coefficients
+from ufl.constantvalue import RealValue, Zero
+from ufl.corealg.map_dag import map_expr_dag
+from ufl.corealg.multifunction import MultiFunction
+
+from firedrake.utils import ScalarType
+from firedrake.vector import Vector
+
+
+class WeightedCoefficientCollector(MultiFunction):
+    def __init__(self):
+        super().__init__()
+        self.weighted_coefficients = []
+
+    def sum(self, o):
+        for op in o.ufl_operands:
+            self(op)
+
+    def product(self, o):
+        try:
+            weight, = filter(
+                lambda op: isinstance(op, (RealValue, Zero)), o.ufl_operands
+            )
+            coeff, = filter(lambda op: isinstance(op, ufl.Coefficient), o.ufl_operands)
+        except ValueError:
+            raise ValueError("Only products of scalars and coefficients allowed")
+        self.weighted_coefficients.append((weight, coeff))
+
+    def coefficient(self, o):
+        self.weighted_coefficients.append((1, o))
+
+    def invalid(self, o):
+        raise ValueError(f"{o} is not a valid component of a weighted sum of coefficients")
+
+    expr = invalid
+
+
+class Assigner:
+    symbol = "="
+
+    def __init__(self, assignee, expression, subset=None):
+        if isinstance(expression, Vector):
+            expression = expression.function
+
+        if not all(
+            c.function_space() == assignee.function_space()
+            for c in extract_coefficients(expression)
+        ):
+            raise ValueError
+
+        self._assignee = assignee
+        self._expression = expression
+        self._subset = subset
+
+    def __str__(self):
+        return f"{self._assignee} {self.symbol} {self._expression}"
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._assignee!r}, {self._expression!r})"
+
+    def assign(self):
+        if self._subset:
+            indices = self._subset.indices
+            self._assignee.dat.data[indices] = np.dot(self.weights, self.coefficient_data[indices])
+        else:
+            # if there is no subset then we can also assign halo values to avoid a later halo exchange
+            self._assignee.dat.data_with_halos[...] = np.dot(self.weights, self.coefficient_data_with_halos)
+            self._assignee.dat.halo_valid = True
+
+    @cached_property
+    def weights(self):
+        return np.array(self._weighted_coefficients[0], dtype=ScalarType)
+
+    @cached_property
+    def coefficients(self):
+        return self._weighted_coefficients[1]
+
+    @cached_property
```

Comment:
This caching is unsafe if someone ever holds onto an `Assigner` object and runs it in a loop where the rvalue coefficients are modified elsewise.